Feature/crm system show all tasks

keydmitrij opened the pull request.

keydmitrij committed:
Убрал лишнее из App.tsx, добавил router и корневую страницу

keydmitrij committed:
Добавил css reset

keydmitrij committed:
Добавил название сайта и иконку

keydmitrij committed:
Добавил компонент TodoItem и стили к нему

keydmitrij committed:
Убрал max-width

Из-за нестинга этот парам сужал туду айтем до безобразия

keydmitrij committed:
Создал компонент с туду листом

В компоненте есть некрасивый тайпинг с ! оператором на data типе, но мне
показалось, что это уместно, тк если бы респонса не было, это было бы
обработано еще в загрузчике.

keydmitrij committed:
Добавил минимальную ширину для листа и зазор между задачами

keydmitrij committed:
Сделал возможность добавлять таску

keydmitrij committed:
Добавил валидацию на инпут создания задачи

keydmitrij committed:
Добавил уведомление об успешном создании таски

keydmitrij committed:
Переписал фетчинг в компоненте TodoList на вариант без роутера

keydmitrij committed:
Добавил обработку ошибок при загрузке туду листа

keydmitrij committed:
Убрал обработку ошибок тк работала через жопу и вынес апи колл в модуль

keydmitrij committed:
Добавил сообщение о загрузке туду листа во время фетчинга с бека

keydmitrij committed:
Переписал AddTodo без роутера и привинтил изменение стейта TodoList

keydmitrij committed:
Добавил к компоненту TodoItem кнопку удаления задачи

keydmitrij committed:
Написал апишку для удаления таски

keydmitrij committed:
Сделал фичу с удалением таски

keydmitrij committed:
Добавил автоапдейт туду листа с бека

keydmitrij committed:
Сделал очистку инпута после добавления таски

keydmitrij committed:
Добавил стили для ховера на кнопку добавления таски

keydmitrij committed:
Написал апишку для изменения таски

keydmitrij committed:
Сделал функционал изменения статуса таски

keydmitrij committed:
Добавил кнопку изменения таски и поменял цвет у иконок

keydmitrij committed:
Убрал неиспользуемые классы с иконок

keydmitrij committed:
Добавил возможность сейвить измененную таску по клику на иконку

keydmitrij committed:
Изменил иконку при сохранении имени таски и поменял все png на svg

keydmitrij committed:
Пофиксил стили иконок, теперь не скачут при изменении имени таски

keydmitrij committed:
Добавил функционал фильтра по статусу таски

keydmitrij committed:
Добавил зависимость статус в useEffect

Теперь авторефреш не сбрасывает статус

keydmitrij committed:
Добавил стиля навигации по фильтрам

keydmitrij committed:
Добавил кол-во тасок в списке в навигации

keydmitrij committed:
Исправил баг с подсчетом тасок по статусу

In work и completed были перепутаны местами, сейчас все как надо

keydmitrij committed:
Изменил положение сообщения о загрузке листа

Сделал относительное положение чтобы айтемы не скакали при загрузке

keydmitrij committed:
Добавил кнопку отмены изменения таски

keydmitrij committed:
Добавил условие для вывода ошибки

Теперь ошибка не вылезает если поле было изменено но юзер удалил
все символы

keydmitrij committed:
Добавил локальное изменение стейта после добавления тудушки

Инста отклик, данные подгружаются уже после

keydmitrij committed:
Добавил отступ внизу листа

Теперь не касается границы экрана

keydmitrij committed:
Сделал сначально локальное удаление таски затем отправку реквеста

keydmitrij committed:
Добавил враппер button вокруг иконок взаимодействия с тасками

Должно избавить от иногда не работающего клик ивента

keydmitrij committed:
Сделал так что измененеие таски идет сначала локально

keydmitrij committed:
Сделал локал ферст изменение статуса таски

keydmitrij committed:
Сделал локал ферст изменение статуса

keydmitrij committed:
Убрал задержку изменения кол-ва тасок по фильтру

keydmitrij committed:
Запретил сейв изменения таски если есть ошибка валидации

keydmitrij committed:
Фикс бага с эдитом выполненной задачи и убрал кнопку эдита на выполненной

keydmitrij committed:
Пофиксил баг с изменением идникации кол-ва тасок

keydmitrij committed:
Превратил функцию в IIFE для краткого кода

keydmitrij committed:
Вынес страницу с туду листом из App.tsx в TodoListPage.tsx

keydmitrij committed:
Обернул файл с типами в отдельную папку types

keydmitrij committed:
Убрал полностью локальные фичи и немного рефакторинга

keydmitrij committed:
Обернул запросы к API в try catch

keydmitrij committed:
Немного рефактора и исправил опечатку и лишние ивенты в EditTodo

keydmitrij committed:
Удалил обработку данных которая ничего пока не дает

keydmitrij committed:
Вернул возможность редактировать выполненную таску

seledkov commented on src/types/types.ts:
```diff
@@ -0,0 +1,33 @@
+  isDone?: boolean;
+}
+
+export type todoStatus = "all" | "completed" | "inWork";
```
TodoStatus

seledkov replied:
хорошо / сделал

seledkov replied:
иди на хуй

seledkov replied:
привет объясни что ты имел ввиду написав мой тип с большой буквы

seledkov commented on src/pages/TodoListPage.tsx:
```diff
@@ -0,0 +1,74 @@
+
+const TodoListPage = () => {
+  const [todoList, setTodoList] = useState<Todo[]>([]);
+  const [status, setStatus] = useState("all" as todoStatus);
```
useState<???>("all");

seledkov commented on src/pages/TodoListPage.tsx:
```diff
@@ -0,0 +1,74 @@
+const TodoListPage = () => {
+  const [todoList, setTodoList] = useState<Todo[]>([]);
+  const [status, setStatus] = useState("all" as todoStatus);
+  const [tasksCounter, setTasksCounter] = useState({
```
useState<???

seledkov commented on src/pages/TodoListPage.tsx:
```diff
@@ -0,0 +1,74 @@
+
+  const updateTaskCounter = async () => {
+    const taskCounters = await fetchTodoList("all");
+    if (taskCounters instanceof Error) return;
```
if (taskCounters instanceof Error) {
   return;
   }

seledkov commented on src/pages/TodoListPage.tsx:
```diff
@@ -0,0 +1,74 @@
+  const updateTaskCounter = async () => {
+    const taskCounters = await fetchTodoList("all");
+    if (taskCounters instanceof Error) return;
+    const allTasksCounter = taskCounters!.info!.all;
```
! утверждение не юзай НИКОГДА тк это приводит тк багам

seledkov commented on src/pages/TodoListPage.tsx:
```diff
@@ -0,0 +1,74 @@
+
+    if (fetchedData instanceof Error === false) {
+      setTodoList(fetchedData.data);
+      updateTaskCounter();
```
удалить отдельный запрос чтобы не было гонки транзакций или проще говоря в разных запросах могут придти разные данные даже если ты "одновременно" запросы сделал

seledkov commented on src/pages/TodoListPage.tsx:
```diff
@@ -0,0 +1,74 @@
+    }
+  };
+
+  useEffect(() => {
```
удалить потому что в тз таокго нет

seledkov commented on src/pages/TodoListPage.tsx:
```diff
@@ -0,0 +1,74 @@
+    <>
+      <AddTodo
+        status={status}
+        updateTasks={updateTasks}
```
updateTasks={()=>updateTasks(status)}

вариант 2 не самый лучший но какой есть

seledkov commented on src/pages/TodoListPage.tsx:
```diff
@@ -0,0 +1,74 @@
+    }));
+  };
+
+  const updateTasks = async (status: todoStatus) => {
```
const updateTasks = async () => {
    const fetchedData = await fetchTodoList(status); // статус из юзстейта чтобы не передавать статус во все месте где нужна функция апдейтаскс

seledkov commented on src/components/AddTodo.tsx:
```diff
@@ -0,0 +1,63 @@
+
+  const handleSubmit = async (event: React.SyntheticEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (titleHasError || titleValue.length === 0) return;
```
if (titleHasError ) return;

вся валидация идет в извалидленгтх дублирование проверок если требуется то значит неправильно реализована валидация

seledkov commented on src/components/AddTodo.tsx:
```diff
@@ -0,0 +1,63 @@
+    setIsUploadingTask(true);
+
+    resetTitle();
+    await createTodoItem(titleValue);
```
await createTodoItem(titleValue);
resetTitle();

seledkov commented on src/components/AddTodo.tsx:
```diff
@@ -0,0 +1,63 @@
+    event.preventDefault();
+    if (titleHasError || titleValue.length === 0) return;
+    setIsUploadingTask(true);
+
```
трайкетч

seledkov commented on src/api/todo.ts:
```diff
@@ -0,0 +1,111 @@
+    }
+    return resData;
+  } catch (error) {
+    return error as Error;
```
throw error

seledkov commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,92 @@
+
+  return (
+    <li className={classes["todo-item"]}>
+      <img
```
сделать не имаджем / или добавить инпут чекбокс

seledkov commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,92 @@
+      )}
+
+      {isEditing && (
+        <EditTodo
```
развернуть компонент обратно тк мы ничего не выигрываем от его выноса

seledkov commented on src/components/TodoItem.tsx:
```diff
@@ -0,0 +1,92 @@
+  };
+
+  const handleEditButton = () => {
+    setIsEditing((isEditing) => !isEditing);
```
лучше юзать туглеры только для чекбоксов а для таких кейсов делай явные хандлеры стартедит ендедит

keydmitrij committed:
Исправил опечатку в названии типа TodoStatus и поменял as TodoStatus на <TodoStatus>

keydmitrij committed:
Удалил авторефреш туду листа

keydmitrij committed:
Удалил двойной запрос приводящий к гонке транзакций

keydmitrij committed:
Убрал status из зависимостей useEffect

Из-за этого useEffect срабатывает только при маунте и не дает ему отправлять
доп запрос при изменении статуса, соответственно вместо 2 запросов - 1

keydmitrij committed:
Переписал updateTasks покороче

keydmitrij committed:
На всякий разделил установку стейта, вдруг инфо не придет

keydmitrij committed:
Прокинул функцию обновления сразу с аргументом

Не нужно прокидывать второй проп и вызывать функцию с ним в другом компоненте

keydmitrij committed:
Удалил лишнй импорт

keydmitrij committed:
Обернул добавление таски в трайкетч

keydmitrij committed:
Дал функции валидации более подходящее имя

keydmitrij committed:
Изменил валидацию и перенес логику хука в компонент

keydmitrij committed:
Переписал updateTasks

Теперь не требуется пропихивать status в пропсах, он берется из стейта

keydmitrij committed:
Удалил вынесенный EditTodo и обернул кнопку статуса в button

keydmitrij committed:
Навел порядок в файле стилей

keydmitrij committed:
Выключил кнопку добавления таски на время загрузки на сервер

keydmitrij committed:
Заменил return на throw ошибок и обычные строки на литералы в URL

keydmitrij committed:
Добавил обработку ошибки при неудаче получения туду листа

keydmitrij committed:
Обернул запросы в трай кетч и добавил разные статусы ошибок

keydmitrij committed:
Пофиксил показ валидации создания таски

keydmitrij committed:
Сделал сброс черновика после отмены изменений
